primitives/trie/src/cache.rs

```diff
@@ -0,0 +1,828 @@
+	/// Keeps track of all the trie nodes accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache.
+	shared_node_cache_access: Mutex<HashSet<H::Out>>,
+	/// The local cache for the values.
+	value_cache: Mutex<IntMap<u64, CachedValue<H::Out>>>,
+	/// Keeps track of all values accessed in the shared cache.
+	///
+	/// This will be used to ensure that these nodes are brought to the front of the lru when this
+	/// local instance is merged back to the shared cache.
+	shared_value_cache_access: Mutex<IntSet<u64>>,
+}
+
+impl<H: Hasher> LocalTrieCache<H> {
+	/// Return self as a [`TrieDB`](trie_db::TrieDB) compatible cache.
+	///
+	/// The given `storage_root` needs to be the storage root of the trie this cache is used for.
+	pub fn as_trie_db_cache<'a>(&'a self, storage_root: H::Out) -> TrieCache<'a, H> {
+		let value_cache = ValueCache::ForStorageRoot {
+			storage_root,
+			local_value_cache: self.value_cache.lock(),
+			shared_value_cache: self.shared.value_cache.read(),
+			shared_value_cache_access: self.shared_value_cache_access.lock(),
+		};
+
+		TrieCache {
+			shared_node_cache: self.shared.node_cache.read(),
+			local_cache: self.node_cache.lock(),
+			value_cache,
+			shared_node_cache_access: self.shared_node_cache_access.lock(),
+		}
+	}
+
+	/// Return self as [`TrieDBMut`](trie_db::TrieDBMut) compatible cache.
+	///
+	/// After finishing all operations with [`TrieDBMut`](trie_db::TrieDBMut) and having obtained
+	/// the new storage root, [`TrieCache::merge_into`] should be called to update this local
+	/// cache instance. If the function is not called, cached data is just thrown away and not
+	/// propagated to the shared cache. So, accessing these new items will be slower, but nothing
+	/// would break because of this.
+	pub fn as_trie_db_mut_cache<'a>(&'a self) -> TrieCache<'a, H> {
+		TrieCache {
+			shared_node_cache: self.shared.node_cache.read(),
+			local_cache: self.node_cache.lock(),
+			value_cache: ValueCache::Fresh(Default::default()),
+			shared_node_cache_access: self.shared_node_cache_access.lock(),
+		}
+	}
+}
+
+impl<H: Hasher> Drop for LocalTrieCache<H> {
+	fn drop(&mut self) {
+		self.shared
+			.node_cache
+			.write()
+			.update(self.node_cache.lock().drain(), self.shared_node_cache_access.lock().drain());
+
+		let mut shared_value_cache = self.shared.value_cache.write();
+		let mut shared_value_cache_access = self.shared_value_cache_access.lock();
+
+		// First write the new values that were cached here locally
+		self.value_cache.lock().drain().for_each(|(k, v)| {
+			shared_value_cache_access.remove(&k);
+			shared_value_cache.put(k, v);
+		});
+		// Then only touch the ones that we have read from the global cache, to bring them
+		// up in the lru.
+		shared_value_cache_access.drain().for_each(|k| {
+			shared_value_cache.get(&k);
+		});
```

Comment:
Hmm... this isn't necessarily a huge problem, but I'd just like to point out (as it's not immediately obvious) that since the `shared_value_cache_access` is a hashset this will essentially bump those keys in the LRU in a semi-random order and not strictly in the order in which they were accessed.

But I think we can ignore this for now. They ideal fix for this would most likely be to replace the LRU with a lock-free one which can be updated as we go; that's totally out of scope of this PR though as the performance is good enough as-is for a v1.